libnetwork/osl/namespace_linux.go

```diff
@@ -555,32 +539,15 @@ func (n *Namespace) Restore(interfaces map[Iface][]IfaceOption, routes []*types.
 	return nil
 }
 
-// Checks whether IPv6 needs to be enabled/disabled on the loopback interface
-func (n *Namespace) checkLoV6() {
-	var (
-		enable = false
-		action = "disable"
-	)
-
-	n.mu.Lock()
-	for _, iface := range n.iFaces {
-		if iface.AddressIPv6() != nil {
-			enable = true
-			action = "enable"
-			break
-		}
-	}
-	n.mu.Unlock()
-
-	if n.loV6Enabled == enable {
-		return
-	}
-
-	if err := setIPv6(n.path, "lo", enable); err != nil {
-		log.G(context.TODO()).Warnf("Failed to %s IPv6 on loopback interface on network namespace %q: %v", action, n.path, err)
+// IPv6LoEnabled checks whether the loopback interface has an IPv6 address ('::1'
+// is assigned by the kernel if IPv6 is enabled).
+func (n *Namespace) IPv6LoEnabled() bool {
+	iface, err := n.nlHandle.LinkByName("lo")
+	if err != nil {
+		return false
 	}
-
-	n.loV6Enabled = enable
+	addrs, err := n.nlHandle.AddrList(iface, nl.FAMILY_V6)
+	return err == nil && len(addrs) > 0
```

Comment:
I wonder if we should actually check that `addrs` contains `::1`. While a fairly safe assumption today with runC and a vanilla kernel, all bets are off with custom kernels and alternative runtimes. For example, a runtime could support some custom OCI annotation which is analogous to `ip addr del ::1 dev lo`.

Reply:
I've been thinking of the absence of `::1` as a proxy for "no IPv6 support in this container" ... which I think is a fairly safe assumption, `::1` is a [required address](https://datatracker.ietf.org/doc/html/rfc4291#section-2.8) for an IPv6 host, and Linux's implementation of that is to add it to `lo`.

Only removal of `::1 localhost` from `/etc/hosts` is specific to `::1`.

Removal of `fe00::0 ip6-localnet` and the other defaults, or the IPv6 address of the container itself and (in later changes), not allocating IPv6 addresses, and generating an appropriate `resolv.conf` are based on deciding that the container has no IPv6 support - not specifically that it has no `::1` on `lo`.

So, checking for the presence of any IPv6 address on `lo`, rather than specifically for `::1` is equivalent for that purpose. If there is an IPv6 address it's probably `::1` but in any case, if it's there, the container supports IPv6. (And, if it's a standards-compliant implementation, `::1` will work somehow.)

If lack of a `::1` on `lo`, in a container that supports IPv6, is a real-world problem... I'll need a better test for "IPv6 enabled". Perhaps trying to assign an IPv6 address and looking for an error - but I think that's probably over-the-top.